haystack/pipelines/config.py

```diff
@@ -67,16 +66,33 @@ def get_component_definitions(
                                          `_` sign must be used to specify nested hierarchical properties.
     """
     component_definitions = {}  # definitions of each component from the YAML.
-    raw_component_definitions = copy.deepcopy(pipeline_config["components"])
-    for component_definition in raw_component_definitions:
+    for component_definition in pipeline_config["components"]:
+        name = component_definition["name"]
+        component_definitions[name] = {key: value for key, value in component_definition.items() if key != "name"}
```

Comment:
Let's rename `component_definition` to `raw_component_definition`, set `component_definition` to this and assign it to `component_definitions[name]`. Otherwise the `_overwrite_with_env_variable` would need to work on `component_definitions[name]` to not change the original one.
```suggestion
        component_definition = {key: value for key, value in raw_component_definition.items() if key != "name"}
        component_definitions[name] = component_definition
```

Reply:
Hey, I don't think my code will overwrite `_component_config`, so I wrote a test to check and it seems to be passing with the current version. Have a look if the test is actually testing the bug you are highlighting.

Reply:
I added a comprehensive test checking whether `_component_config` or `get_config()` are in any way affected, but it doesn't seem like. The original `pipeline_config`, in fact, is assembled by `get_config()`, so I don't think there's a reason for concern. Actually I couldn't find any point where the changes applied by `_overwrite_with_env_variable` were spilling out. Did I understand your comment correctly? Have a look at the new test and let me know if I forgot something.

Reply:
Yes, in the test it just doesn't have an effect because you always create a new config via `get_config()`. Once you take the same instance (take the one before deepcopy of course), you'll find it changed. This is problematic as we cannot assume that pipeline configs always come from `get_config` and are not used afterwards anymore. The config could be loaded and processed in any thinkable ways, having a call to `get_component_definitions` in between. So let's ensure it's still the same afterwards

Reply:
Ok I see your point now, but I don't know how we can achieve it without deepcopying `pipeline_config`. The renaming you suggest does not achieve that goal. I add a test for this specific case.

I think the options are:
- We keep the deepcopy and set #2568 as wontfix (even though this might bite us later), or
- Document this behavior and make `get_components_config` mutate `pipeline_config` (even though this, as well, might bite us later).

I will think about this more and see if I can come up with a way around it. Thanks for bringing it up btw, I didn't fully get the consequences of applying `_overwrite_with_env_vars` here.

Reply:
Ok I came up with a conditional copy that might be solving the issue. Basically `pipeline_config` will be deepcopied only if there are env vars values to apply to the pipeline config, and when the copy fails it returns a meaningful error message. Do you think this could work?